File: content-solutions/backtracking/number-cards.py
def solution(cards, k):
    n = len(cards)
    used = [False] * n
    path = []
    made = set()

    def arrange():
        if len(path) == k:
            if path[0] != 0:
                made.add("".join(map(str, path)))
            return
        for i in range(n):
            if not used[i]:
                used[i] = True
                path.append(cards[i])
                arrange()
                path.pop()
                used[i] = False

    arrange()
    return len(made)

File: content-solutions/backtracking/number-cards.js
function solution(cards, k) {
  const n = cards.length;
  const used = new Array(n).fill(false);
  const path = [];
  const made = new Set();
  function arrange() {
    if (path.length === k) {
      if (path[0] !== 0) made.add(path.join(""));
      return;
    }
    for (let i = 0; i < n; i += 1) {
      if (!used[i]) {
        used[i] = true;
        path.push(cards[i]);
        arrange();
        path.pop();
        used[i] = false;
      }
    }
  }
  arrange();
  return made.size;
}
